Inference using Dev Containers

lasnown opened the pull request.

lasnown committed:
Changes to the existing files

lasnown committed:
Added files for model inference using Dev Container

lasnown committed:
Updated README.md

myadryshnikova commented on model_inference/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM python:3.11
```
Try using python slim images. They will most likely fit your task as well as this one, but will weigh much less and be faster to build since it only has basic python customization.
3.11-slim should work for you.

myadryshnikova commented on model_inference/.devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,18 @@
+      ]
+    },
+	"forwardPorts": [3000],
+	"postCreateCommand": "pip install -r requirements.txt"
```
I think it's better to move this command to Dockerfile. Then this file will become more universal, because you can use it not only for devcontainer, but also for other tasks.

myadryshnikova replied:
That's cool that you found that property, never really used that property to set dependencies.

myadryshnikova commented on model_inference/.devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,18 @@
+      "ms-python.python"
+      ]
+    },
+	"forwardPorts": [3000],
```
I'm not sure you need it in this situation. The ports don't seem to be used in any further way, they don't open access to anything. I have a feeling it can be painlessly removed.

myadryshnikova commented on model_inference/bird_predictor.py:
```diff
@@ -17,12 +17,18 @@
+            file.write(chunk)
+
 def model_upload():
   output = 'bird_recognition.h5'
```
Why is this thing called "output"? If I understand the situation correctly, it is the model weights. Can you rename this variable according to its role? In this situation, I would first think it is the prediction output from the model.

myadryshnikova commented on model_inference/bird_predictor.py:
```diff
@@ -17,12 +17,18 @@
         6: 'Цыпленок'
     }
 
+def download_file_from_url(url, destination_path):
```
Let's try adding typing to all functions.
How, for example, typing might look like in this case:
def download_file_from_url(url: str, destination_path: str) -> None:

lasnown committed:
build: change base image and move dependency installation process

Change base image in Docker file to python3.11-slim. Move dependency installation from devcontainer.json to Dockerfile.

lasnown committed:
refactor(predictor): add typing to functions

lasnown committed:
feat(predictor): implement OOP elements

Added the following classes: Model, which includes all the methods related to the model; Image, which contains info about the image path and the image preprocessing method.

myadryshnikova commented on README.md:
```diff
@@ -16,14 +18,21 @@
 
+#### 2 вариант: Используя расширение Dev Containers для VSC
+
+1. Скачать и распаковать архив с файлами репозитория в локальный каталог.
```
Let's leave it more like "clone the repository with a 

'''bash
git clone _git-repo-url_
'''
"

myadryshnikova commented on model_inference/bird_predictor.py:
```diff
@@ -17,32 +15,49 @@
-    gdown.download(url, output, quiet=False)
-  model = load_model(output)
-  return model
+def download_file_from_url(url:str, destination_path:str) -> None:
```
Let's try adding docstrings to all classes and functions. Example:
```
def add_binary(a, b):
    '''
    Возвращает сумму двух десятичных чисел в двоичном формате.
            Параметры:
                    a (int): первое десятичное целое число
                    b (int): второе десятичное целое число
            Возвращаемое значение:
                    binary_sum (str): двоичная строка суммы a и b
    '''
    binary_sum = bin(a+b)[2:]
    return binary_sum
```


This is needed so that to all functions or classes you can call the __doc__ method and get docstrings. You can try to paste this example somewhere and call it:
```
print(add_binary.__doc__)
```

Let's do the same thing, but only in English.

myadryshnikova commented on model_inference/bird_predictor.py:
```diff
@@ -17,32 +15,49 @@
+    kind=kind[0]
+    return score,kind
+  
+class Image:
```
It's a good idea to have a separate class for image preprocessing. But it seems to me a bit counter-intuitive to call the class exactly Image.
Why: in theory, a class is an object, and its methods are what the object can do. So it's a bit out of sync with my understanding of reality. An image that processes an image? Maybe try something like ImagePreprocessor, which in init-method has pass, and the input of the image conversion function gets the path to it? Please consider this, I could be wrong.

myadryshnikova commented on model_inference/bird_predictor.py:
```diff
@@ -17,32 +15,49 @@
+  def make_predictions(self, img_array:np.ndarray) -> tuple[np.float32, np.int64]:
+    predictions = self.__model.predict(img_array)
+    score = np.max(predictions)
+    kind = np.where(predictions==np.max(predictions))[1]
+    kind=kind[0]
+    return score,kind
```
I have to say, I'm a formatting maniac. I have a hard time seeing "=" and "==" without spaces around them, "," without a space after it. I suggest fixing that too, although it's certainly not a bug or anything like that :)

You can spend some more time and try to customize the linking for this project to check codestyle. Suitable plugin: https://pypi.org/project/pylint/

After that, you can try to write a Github Action that will trigger a linting job on every pull-request and prevent the merge pull-request in the master-branch if the linting is unsuccessful.

lasnown committed:
refactor(predictor): use ImagePreprocessor class instead of Image

lasnown committed:
docs(predictor): add docstrings to functions and classes
style(predictor): apply linting

lasnown committed:
build(requirements): change the dependencies' list

Add pylint and related libraries to the dependencies' list. Remove redundant dependencies.

lasnown committed:
docs: update README

lasnown committed:
docs: update README.md

myadryshnikova commented on README.md:
````diff
@@ -4,33 +4,79 @@
-4. По завершении процесса создания образа использовать команду **docker-compose up**.
+3. Создать образ, используя:
+
+   ```bash
+   docker-compose build
+   ```
+
+4. По завершении процесса создания образа использовать следующую команду для запуска:
+
+   ```bash
+   docker-compose up
+   ```
````
We can also use the command  docker-compose up --build. It's more compact and easier to understand.

myadryshnikova commented on model_inference/bird_predictor.py:
```diff
@@ -17,32 +16,125 @@
+        for chunk in response.iter_content(chunk_size = 128):
+            file.write(chunk)
 
+class BirdRecModel:
```
Maybe BirdRecognitionModel? I think it's an unnecessary reduction.

myadryshnikova commented on model_inference/bird_predictor.py:
```diff
@@ -17,32 +16,125 @@
+        kind = kind[0]
+        return score, kind
+
+class ImagePreprocessor:
```
Let's try to describe one class in one file and import the classes used into the main program. We can create the following folder structure:

model_inference
|----modules
|----|-----|image_preprocessor
|----|-----|-----|image_preprocessor.py
|----|-----|-----|__init__.py
|----main.py

Commit:
conf: update .gitignore

Commit:
docs: update README.md

lasnown committed:
fix: delete Chicken.webp

Commit:
conf: update .gitignore

Commit:
build: update requirements.txt

Commit:
feat: create image processing and model initialization modules

Commit:
feat: add main.py

Commit:
feat: remove redundant bird_predictor.py

Commit:
style: fix linting issues

Commit:
docs: unnoticable change to test actions

Commit:
docs: unnoticable change to test actions

Commit:
docs: unnoticable change to test actions

Commit:
docs: unnoticable change to test actions

Commit:
docs: unnoticable change to test actions

Commit:
checking

lasnown committed:
action: update pylint.yml

lasnown committed:
actions: delete .github/workflows/pylint.yml

lasnown committed:
fix: add new an empty public method to model initialization module

lasnown committed:
fix: add the second public class to image preprocessing module

lasnown committed:
styling: fix linting issues

lasnown committed:
actions: add .yml to the inference branch
